crates/bevy_ecs/src/system/commands/parallel_scope.rs

```diff
@@ -58,12 +58,12 @@ impl<'w, 's> SystemParamFetch<'w, 's> for ParallelCommandsState {
     unsafe fn get_param(
         state: &'s mut Self,
         _: &crate::system::SystemMeta,
-        world: &'w World,
+        world: MaybeUnsafeCell<'w, World>,
```

Comment:
I'm generally on board for the approach in this PR, but having thought about this problem a bit, I think all of this `MaybeUnsafeCell` and `world_access_level` validation stuff is a sign that we're doing this inside-out. If we can never safely call an exclusive system with a `&World` reference, we should probably try to just represent that statically.

To that end, I have [put together an alternative PR](#6083) that adds a new `ExclusiveFunctionSystem`, which implements `System` (backed by `ExclusiveSystemParam` impls and the assumption that `&mut World` is always passed in).

This means that we can leave all of the existing "shared access system function" stuff untouched, and we can piggyback on the existing `System::run` and `System::run_unsafe` functions to ensure that exclusive systems are only run with `System::run` (by panicking on `System::run_unsafe` calls).

This impl is smaller, less invasive, and introduces fewer safety concerns. It comes at the cost of having a hard line between functions with SystemParams and ExclusiveSystemParams, but the impl in this PR _already_ had that constraint. It also means that we have another set of ExclusiveSystemParam tuple impls, but I'm not noticing any difference in compile times.

Reply:
> this `MaybeUnsafeCell` and `world_access_level` validation stuff is a sign that we're doing this inside-out

`MaybeUnsafeCell` is a workaround to `&T -> &UnsafeCell<T>` not being a valid cast. @BoxyUwU suggested that [may change](https://github.com/rust-lang/unsafe-code-guidelines/issues/303).

And it's not inside-out. Having to prove soundness ourselves is the nature of our access model. The "stance" this PR takes is that the model should be complete (able to express any sound access). `world_access_level` is the natural extension to fill that gap. Your PR trades completeness for the static guarantee that users can't write systems with unsound access on this level (queries are still checked at runtime).

> fewer safety concerns

It's a little unfair to say that what's implemented here is less safe. The contract of `SystemParamState` is that params declare their access truthfully. UB can only result from violating that contract. That's still true. And while not statically prevented, unsound `World` access here panics at system construction, which is earlier than unsound query access.

> it comes at the cost of having a hard line between functions with SystemParams and ExclusiveSystemParams the impl in this PR _already_ had that constraint

This PR does not have a hard line exclusive/parallel *trait* split, which is what causes inconsistencies. #6083 wouldn't let a user implement a custom param with an `&mut World` field or use `&mut World` in `ParamSet`.

> less invasive

I'll give you that. Your PR does save custom param implementers from having to change anything.

Reply:
> But the general idea is not inside-out. Having to prove soundness ourselves is the nature of our access model. This PR tries to complete this model so it can express any sound access, and world_access_level is the natural extension to fill that gap.

Within this context, I think it is. SystemParams exist to prove to the compiler that extracting disjoint components / resources from `&World` is sound. This is impossible to do in safe Rust.

`MaybeUnsafeCell` and `world_access_level` exist to ensure that `&mut World` is never used alongside `&World` derived references and that `&World` is never casted to `&mut World`. This is possible to do in safe Rust (and is kind of the selling feature of it).

These are different problem spaces solving different problems. I do think that choosing to merge them under one API is valid, but I think in this case doing that isn't worth the complexity / risk.

> It's a little unfair to say that what's implemented here is less safe. While true, the contract of SystemParamState is that params declare their access truthfully. UB can only result from violating that contract. That's still true.

I don't think this is unfair. It expands the scope of the safety contract, making it harder to implement and verify. I think calling that "less safe" is a very fair assessment.

> And while not statically prevented, unsound World access here panics at system construction, which is still earlier than unsound query access.

_Only_ if as a SystemParam implementer you implement `world_access_level` correctly / only extract `&mut World` from `MaybeUnsafeCell` under the right circumstances.

> This PR does not have a hard line exclusive/parallel trait split, which is what causes arbitrary limits / inconsistencies. E.g. https://github.com/bevyengine/bevy/pull/6083 would prevent a user from implementing a custom param with a &mut World field and from using &mut World in a ParamSet.

Yup I agree that putting `&mut World` in a `ParamSet` would be impossible in #6083, but I think thats actually desirable (as called out in that pr).

Reply:
> It expands the scope of the safety contract, making it harder to implement and verify.

> *Only* if as a `SystemParam` implementer you implement `world_access_level` correctly / only extract `&mut World` from `MaybeUnsafeCell` under the right circumstances.

The same could be said for `ReadOnlySystemParamFetch` and `SystemParamState`.  I'm happy to accept this as an argument about aesthetics (i.e. "let's not make the contract longer"), but I don't really buy the safety angle here. Ultimately, our model has to assume honesty and correctness. We can't verify soundness if one of those contracts is violated anyway.